querymsc.py

```diff
@@ -505,5 +507,124 @@ def query_rain_rate_one_day_1_50(self, session):
+        """
+
+        N = x.shape[0]
+
+        # euler-mascheroni constant
+        euler = 0.5772156649
+
+        if N >= self.min_fit:
+            # create fitted gumbel object to
+            # distribution of extreme values
+            paras = distr.gum.lmom_fit(x)
+
+            # extract L-moments
+            lmoments = distr.gum.lmom(nmom=2, **paras)
+
+            # calculate estimators of gumbel
+            # parameters
+            alpha = lmoments[1]/np.log(2)
+            xi = lmoments[0] - euler*alpha 
+
+            return xi, alpha
+        else:
+            return np.nan, np.nan
+
+    def get_design_value(self, xi, alpha):
+        """Get the design value from explicit expression
+        derived from a right-skewed Gumbel distribution.
+        See methods.pdf for more details. Separate from
+        Gumbel fitting in case user defines custom
+        estimators. 
+        -----------------------------------
+        Args:
+            xi (float): Estimated value of the location parameter
+                (first l-moment)
+            alpha (float): Estimated value of the scale parameter
+                (second l-moment).
+        Returns:
+            design_val (float): design value using Gumbel parameters
+        """
+
+        # return frequency, i.e. inverse of return period 
+        f_r = 1.0/self.return_period
+
+        # simplify long expression
+        simp = (1.0-f_r) + np.exp(-np.exp((xi/alpha)))
+
+        # final expression for design value
+        design_val = xi - alpha*np.log(-np.log(simp))
+
+        return design_val
+
+    def get_fit_transform(self, x):
+        """Extract the design value.
+        -----------------------------
+        Args:  
+            x (pandas Series): Series containing the annual
+            grouped extreme values for a given 
+            station over a range of years.
+            
+        Returns:
+            design value (float): the design value
+            as extracted by get_design_value at
+            the given return period.
+        """
+        xi, alpha = self.fit(x)
+        return self.get_design_value(xi, alpha)
+
+    def fit_transform(self, df, variable='rainfall_rate'):
+        """Take dataframe with data from variables 
+        needed for design value calculation, group into
+        stations based on their station_id, and apply
+        the gumbel fitting to get the design values.
+        ------------------------------------------
+        Args:
+            df (pandas DataFrame): Contains all observations
+                of variable of interest
+            variable (Str): Key for the variable of interest.
+                Default key is rainfall_rate
+        
+        Returns:
+            df_new (pandas DataFrame): Same as input dataframe
+                but with added column containing the design values
+                with suffix "_design_val"
+        """
+        df_new = df.join((df.groupby('station_id')[variable]
```

Comment:
OK, so the `DataFrame` has to have a column named `station_id`? That's important to state in the docstring.

Reply:
Agreed!